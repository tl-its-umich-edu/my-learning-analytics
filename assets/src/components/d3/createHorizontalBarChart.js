import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { margin } from '../../constants/chartConstants'

function createHorizontalBarChart ({ data, width, height, el, tip }) {
  const modifiedMargin = { ...margin, left: 120 }
  const [aWidth, aHeight] = adjustViewport(width, height, modifiedMargin)

  const x = d3.scaleLinear()
    .domain([0, d3.max(data, d => d.data)]).nice()
    .range([modifiedMargin.left, aWidth - modifiedMargin.right])

  const y = d3.scaleBand()
    .domain(data.map(d => d.label))
    .range([aHeight - modifiedMargin.bottom, modifiedMargin.top])
    .padding(0.1)

  const svg = d3.select(el).append('svg')
    .attr('width', aWidth)
    .attr('height', aHeight)

  const bar = svg.selectAll('.bar')
    .data(data).enter()
    .append('rect')
    .attr('class', 'bar')
    .attr('x', d => x(0))
    .attr('width', d => x(d.data) - x(0))
    .attr('y', d => y(d.label))
    .attr('height', y.bandwidth())

  svg.append('g')
    .attr('transform', `translate(0, ${aHeight - modifiedMargin.bottom})`)
    .call(d3.axisBottom(x))

  svg.append('g')
    .attr('transform', `translate(${modifiedMargin.left},0)`)
    .call(d3.axisLeft(y))

  if (tip) {
    svg.call(tip)
    bar
      .on('mouseover', tip.show)
      .on('mouseout', tip.hide)
  }
}

export default createHorizontalBarChart
