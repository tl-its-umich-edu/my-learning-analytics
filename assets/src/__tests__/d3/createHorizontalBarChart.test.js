/* global describe, test, expect */
import createHorizontalBarChart from '../../components/d3/createHorizontalBarChart'
import { horizontalBarChartData } from '../testData/d3TestData'

describe('createHorizontalBarChart', () => {
  test('should build a horizontal bar chart', () => {
    const div = document.createElement('div')
    createHorizontalBarChart({ data: horizontalBarChartData, width: 1000, height: 500, el: div })
    expect(div).toMatchSnapshot()
  })
})
