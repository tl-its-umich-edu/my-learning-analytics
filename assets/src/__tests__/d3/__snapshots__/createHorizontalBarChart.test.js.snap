// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHorizontalBarChart should build a horizontal bar chart 1`] = `
<div>
  <svg
    height="450"
    width="860"
  >
    <rect
      class="bar"
      height="50.70422535211267"
      width="18"
      x="120"
      y="363.66197183098586"
    />
    <rect
      class="bar"
      height="50.70422535211267"
      width="54"
      x="120"
      y="307.32394366197184"
    />
    <rect
      class="bar"
      height="50.70422535211267"
      width="180"
      x="120"
      y="250.98591549295773"
    />
    <rect
      class="bar"
      height="50.70422535211267"
      width="684"
      x="120"
      y="194.64788732394365"
    />
    <rect
      class="bar"
      height="50.70422535211267"
      width="126"
      x="120"
      y="138.30985915492957"
    />
    <rect
      class="bar"
      height="50.70422535211267"
      width="72"
      x="120"
      y="81.97183098591549"
    />
    <rect
      class="bar"
      height="50.70422535211267"
      width="162"
      x="120"
      y="25.633802816901408"
    />
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 420)"
    >
      <path
        class="domain"
        d="M120.5,6V0.5H840.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(120.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(210.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(300.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(390.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          15
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(480.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(570.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          25
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(660.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          30
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(750.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          35
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(840.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          40
        </text>
      </g>
    </g>
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(120,0)"
    >
      <path
        class="domain"
        d="M-6,420.5H0.5V20.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,389.0140845070422)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Ethiopia
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,332.67605633802816)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Japan
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,276.3380281690141)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Sweden
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,220)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Canada
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,163.66197183098592)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          China
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,107.32394366197182)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          France
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,50.985915492957744)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          South Korea
        </text>
      </g>
    </g>
  </svg>
</div>
`;
